streams/streams.md

```diff
@@ -0,0 +1,126 @@
+
+Let's find a better solution for this.
+
+## Better approach
+Node is asynchronous in handling I/O-bound tasks. Callbacks are often used for handling these asynchronous tasks, but Node has another way of handling this: [Streams](https://nodejs.org/api/stream.html). Streams come from Unix, where you might know them from using the `|` (pipe) in your shell. You are able to pipe data through multiple stages to its final destination, where each stage can transform the data. There are some excellent articles [written](https://github.com/substack/stream-handbook) [about](http://maxogden.com/node-streams.html) [Streams](https://www.sitepoint.com/basics-node-js-streams/) in Node.
+
+Node's `fs` module has a method `createReadStream`, to create a stream from a file. This stream does not buffer the entire contents of the file, but it emits an event as soon as a chunk of data is available. This is useful for us, since we'd only like to process the first two lines of a file and ignore the rest.
+
+The streams approach of our code would be:
+```js
+const fs = require('fs');
+const async = require('async');
+const path = require('path');
+const highland = require('highland');
+
+let latest = '20160128_192500';
+let validFiles = [];
+
+// Read all files in ./incrementals folder
+fs.readdir('./incrementals', (e, files) => {
+  async.eachSeries(files.sort(), (file, cb) => {
+    const fileStream = fs.createReadStream(path.join('incrementals', file));
+
+    highland(fileStream)
+      .split()  // split file in lines
+      .take(2)  // only the first two lines are interesting for us
+      .toArray((lines) => {
+        const previous = lines[1].split(' ').pop();
+
+        if (latest && latest !== previous) {
+          cb();
+          return;
+        }
+
+        latest = lines[0].split(' ').pop();
+        validFiles.push(file);
+
+        cb();
+      });
+  }, () => {
+    console.log(validFiles);  // => Prints a list of all valid files
+  });
+});
+```
+
+The thing that has changed is how a single file is processed. Instead of using `fs.readFile`, we now generate a filestream with `fs.createReadStream`. We pass this stream to [Highland](http://highlandjs.org/). Highland is a library that provides useful functions for handling streams and for transforming and iterating data on streams.
+
+Highland transforms the contents of the incremental file as follows:
+- [`split()`](http://highlandjs.org/#split) splits the file by line. Each line that is abstracted is being put on the stream again.
+- [`take(2)`](http://highlandjs.org/#take) creates a new stream with 2 elements: the first two lines of the increment file
+- [`toArray()`](http://highlandjs.org/#toArray) casts this stream to an Array so we can work with it in its callback function
```

Comment:
Maybe explain the flowing mode of streams a bit here too, `toArray()` starts drawing from its input stream until it has enough data, putting all previous streams from paused into flowing mode. Might be nice to talk about lazy evaluation here too, though the code sample doesn't really show that.